src/Microsoft.AspNetCore.Mvc.Core/Routing/MvcEndpointFinder.cs

```diff
@@ -0,0 +1,254 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    internal class UrlPageContext
+    {
+        public string PageName { get; set; }
+
+        public string PageHandlerName { get; set; }
+
+        public object Values { get; set; }
+
+        public string Protocol { get; set; }
+
+        public string Host { get; set; }
+
+        public string Fragment { get; set; }
+    }
+
+    internal class MvcAddress : Address
+    {
+        public ActionContext CurrentActionContext { get; set; }
+
+        public ActionDescriptor TargetActionDescriptor { get; set; }
+
+        public string TargetActionName { get; set; }
+
+        public string TargetControllerName { get; set; }
+
+        public string TargetPageName { get; set; }
+
+        public string TargetHandlerName { get; set; }
+    }
+
+    internal class MvcEndpointFinder : IEndpointFinder
+    {
+        private readonly IEnumerable<Endpoint> Empty = Enumerable.Empty<Endpoint>();
+        private readonly CompositeEndpointDataSource _endpointDatasource;
+        private readonly IActionDescriptorCollectionProvider _actionDescriptorCollectionProvider;
+        private readonly ILogger<MvcEndpointFinder> _logger;
+        private readonly IEnumerable<MatcherEndpoint> _matcherEndpoints;
+
+        public MvcEndpointFinder(
+            CompositeEndpointDataSource endpointDataSource,
+            IActionDescriptorCollectionProvider actionDescriptorCollectionProvider,
+            ILogger<MvcEndpointFinder> logger)
+        {
+            _endpointDatasource = endpointDataSource;
+            _actionDescriptorCollectionProvider = actionDescriptorCollectionProvider;
+            _logger = logger;
+
+            _matcherEndpoints = _endpointDatasource.Endpoints.OfType<MatcherEndpoint>();
+        }
+
+        public IEnumerable<Endpoint> FindEndpoints(Address address)
```

Comment:
The basic idea is to use as much information from MVC as possible to find endpoints _before_ expanding the templates with the route values. The idea is that this finding endpoints step before expansion can improve perf compared to current approach of expanding each template one by one in the table until we can generate a link